snuba/consumers/consumer.py

```diff
@@ -41,11 +41,21 @@
 from snuba.utils.metrics import MetricsBackend
 from snuba.utils.metrics.wrapper import MetricsWrapper
 from snuba.utils.streams.configuration_builder import build_kafka_producer_configuration
-from snuba.writer import BatchWriter, MockBatchWriter
+from snuba.writer import (
+    BatchWriterEncoderWrapper,
+    MockBatchWriter,
+    MockEncoder,
+    WriterTableRow,
+)
 
 logger = logging.getLogger("snuba.consumer")
 
 
+class WriterTableRowBatch(NamedTuple):
```

Comment:
I think you will need the same `__reduce_ex__` implementation you have in JSONRowInsertBatch. This is (maybe was) needed for the multi process consumer that needs to pickle rows generate by the processor method that runs in the subprocess and transfer them back to the main process via shared memory.

Reply:
I think this comment actually reveals why the consumer was serializing to the insert format and those concerns weren’t separated. I bet whoever did that implementation was trying to avoid the flow of:

1. serialize on process 1 to send data to process 2 
2. deserialize the message on process 2
3. re-serialize the message for process 2 to insert

By serializing to JSON, process 2 can skip a couple steps. My intuition is that this efficiency doesn’t buy us much (it could reduce latency but I don't think these writes are extremely latency-sensitive). Do you want to disagree, though? It’s kind of fundamental to this approach

Reply:
Ah yes, definitively we cannot afford to do the JSON serialization before sending to the subprocess. If we did that will be an instant production incident.

Reply:
Spoke offline: we're rejecting this approach